### ISSUE TITLE
False positive on DL3066 - when user is created as part of the Dockerfile

### ISSUE THREAD

**Issue by SimonTreavis** [hadolint/hadolint#1240]
Hi,
We are seeing what we belive is a false positive on DL3066, when we create the docker user as part of the docker build.

Using hadolint 2.15.1 on the following sample Dockerfile, DL3066 is thrown and causes an exit 1 on an info level message.
```shell
$ hadolint --version
Haskell Dockerfile Linter 2.15.1
$ hadolint Dockerfile 
Dockerfile:13 DL3066 info: Non-numeric user-id may not be resolvable by host system
$ echo $?
1
```
The Dockerfile
```dockerfile
FROM ubuntu:26.04

# Create a non-root user with a fixed UID/GID
RUN groupadd -g 10001 appgroup \
 && useradd -u 10001 -g appgroup -m -s /bin/bash appuser

# Set ownership of application files
COPY --chown=appuser:appgroup . /app

WORKDIR /app

# Run subsequent instructions and the container as this user
USER appuser

CMD ["sleep", "60"]
``` 

Example of running the build and showing the user has the requested id and name

```shell
$ docker build . -t sleepy_container
[+] Building 0.6s (9/9) FINISHED                                                                                                                                                                                            docker:default
 => [internal] load build definition from Dockerfile                                                                                                                                                                                  0.0s
 => => transferring dockerfile: 437B                                                                                                                                                                                                  0.0s
 => [internal] load metadata for docker.io/library/ubuntu:26.04                                                                                                                                                                       0.6s
 => [internal] load .dockerignore                                                                                                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                                                                                                       0.0s
 => [1/4] FROM docker.io/library/ubuntu:26.04@sha256:678c6550cc43645e08669028bc177f50be4e7c5b8cca677067b1914d4afc7a03                                                                                                                 0.0s
 => [internal] load build context                                                                                                                                                                                                     0.0s
 => => transferring context: 90B                                                                                                                                                                                                      0.0s
 => CACHED [2/4] RUN groupadd -g 10001 appgroup  && useradd -u 10001 -g appgroup -m -s /bin/bash appuser                                                                                                                              0.0s
 => CACHED [3/4] COPY --chown=appuser:appgroup . /app                                                                                                                                                                                 0.0s
 => CACHED [4/4] WORKDIR /app                                                                                                                                                                                                         0.0s
 => exporting to image                                                                                                                                                                                                                0.0s
 => => exporting layers                                                                                                                                                                                                               0.0s
 => => writing image sha256:77e6c4521b051fe31a254351e4356a88e76fff040f6c59e1004e683c842232eb                                                                                                                                          0.0s
 => => naming to docker.io/library/sleepy_container                                                                                                                                                                                   0.0s
$ docker run -it --rm --entrypoint=/bin/bash sleepy_container 
appuser@aca93124c63a:/app$ id
uid=10001(appuser) gid=10001(appgroup) groups=10001(appgroup)
appuser@aca93124c63a:/app$ 
```

**Comment by kingcrunch** [comment]
I have something similar, but thats even more "false positive"

```
ENV USER_UID=1000 \
    USER_GID=1000

// ...
USER $USER_UID:$USER_GID
```

**Comment by Lunkentuss** [comment]
Could you please elaborate why this is a false positive? In https://github.com/hadolint/hadolint/wiki/DL3066 it states:

> In Kubernetes, when runAsNonRoot: true is set, the kubelet cannot resolve a symbolic username from the container image (no access to /etc/passwd). This causes pod startup failure with an error like:
  
> Error: container has runAsNonRoot and image has non-numeric user (app), cannot verify user is non-root

Meaning this rule is mainly for running the image in Kubernetes. Your description runs it with docker but this has nothing to do with the rule. I tried to run your example image with the following script:


```bash
eval $(minikube docker-env)
docker build -t lazy .

kubectl apply -f - << EOF
apiVersion: v1
kind: Pod
metadata:
  name: lazy
spec:
  containers:
  - name: lazy
    image: docker.io/library/lazy:latest
    imagePullPolicy: Never
    ports:
    - containerPort: 80
    securityContext:
      runAsNonRoot: true
EOF
```

which yields

> container has runAsNonRoot and image has non-numeric user (appuser),
   cannot verify user is non-root (pod: "lazy_default(d12d6169-ce57-47d9-8982-4a3f4c2691f0)"

**Comment by SimonTreavis** [comment]
Thanks, I had missed the k8s part as the fail was in our CI/CD pipeline before we get anywhere near k8s. Our securityContexts provide `runAsUser` also so we do not have the failure you rightly pointed out.

If anyone else comes across this - the fix to keep hadolint from raising DL3066 is to add the following to the Dockerfile.

Passing the USERID as an arg to useradd with -l flag to stop another warning

```Dockerfile
FROM ubuntu:26.04
ARG USERID=10001
# Create a non-root user with a fixed UID/GID
RUN groupadd -g 10001 appgroup \
 && useradd -l -u ${USERID} -g appgroup -m -s /bin/bash appuser

# Set ownership of application files
COPY --chown=appuser:appgroup . /app

WORKDIR /app

# Run subsequent instructions and the container as this user
USER appuser

CMD ["sleep", "60"]
```

Thanks for the help
